Issue 846: Fix parse issue with queries that use "UNION ALL" or "UNION DISTINCT" and lack "FROM" clauses

barrywhart opened the pull request:
Resolves #846

barrywhart commented on src/sqlfluff/core/dialects/dialect_ansi.py:
```diff
@@ -983,7 +983,16 @@ class SelectClauseSegment(BaseSegment):
+        terminator=OneOf(
+            "FROM",
+            "LIMIT",
+            # :HACK: This Sequence("UNION", ... stuff should not be needed,
```
This is a hacky fix, but it works. Thoughts?

NiallRees replied:
Hey @barrywhart would you mind pushing this to a branch on the sqlfluff repo? I'd like to clone it and take a look

alanmcruickshank reviewed (changes requested):
Right thinking, but I think we can avoid the hack. I think the issue here is that we need to extend `SetOperatorSegment` to include distinct. *most* of what you've added is already in `SetOperatorSegment`, it's just missing `DISTINCT`.

barrywhart commented:
@alanmcruickshank: Can you clarify what you have in mind? AFAICT, the extra line I added to `terminator` was already present in `SetOperatorSegment` -- I simply copied and added it to `terminator`. I definitely agree that it's cleaner to avoid the hacky change I made; I'm simply not understanding the alternative you're suggesting. Thanks!

alanmcruickshank commented:
@barrywhart - found the bug in the parsing logic. You were right, it was a sign of something very very deep in the simple matching heuristic. I think I've found it though and it should be fixed with my most recent commit.
